### PR TITLE
In {box,text}-shadow so blur-radius does not accept negative values

### LINKED ISSUE #15490: The third length in {box,text}-shadow and drop-shadow function should not accept negative value
Spec: https://drafts.csswg.org/css-backgrounds-3/#the-box-shadow

Relevant code: https://github.com/servo/servo/blob/master/components/style/values/specified/mod.rs#L608-L666

`box-shadow` can accept up to four length values, and three of them can accept negative values, while the third value, which is blur radius, should not accept negative value.

Test code:
```html
<!DOCTYPE HTML>
<div style="box-shadow: 1px 1px -1px"></div>
<script>
alert(`decl: ${document.querySelector('div').style.boxShadow}`);
</script>
```

(Note that the serialization result doesn't make sense anyway, which is #15203.)

### PR DESCRIPTION
**Pull request by karan1276**
Change code for serialization for {box,text}-shadow so blur-radius can be parsed as non-negavite

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #15490 (github issue number if applicable).

- [x] Write tests

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15611)

### CODE REVIEW

**Comment by karan1276** [comment]
found some errors. fixing this in some time

**Comment by karan1276** [comment]
My apologies. I think i pushed an empty branch. Ill make those changes again and commit back

**Comment by karan1276** [comment]
Ready for review

File: components/style/values/specified/mod.rs [hunk 1]
```diff
@@ -605,7 +605,7 @@ impl Shadow {
     // disable_spread_and_inset is for filter: drop-shadow(...)
     #[allow(missing_docs)]
     pub fn parse(context:  &ParserContext, input: &mut Parser, disable_spread_and_inset: bool) -> Result<Shadow, ()> {
-        let length_count = if disable_spread_and_inset { 3 } else { 4 };
+        //let length_count = if disable_spread_and_inset { 3 } else { 4 };
```
Comment (upsuper) [inline components/style/values/specified/mod.rs]: Please just remove this line.

File: components/style/values/specified/mod.rs [hunk 2]
```diff
@@ -621,19 +621,15 @@ impl Shadow {
...
+                    lengths[1] = try!(Length::parse(context, input));
+
+                    if let Ok(value) = input.try(|i| Length::parse_non_negative(i)) {
+                        lengths[2] = value;
+
                         if let Ok(value) = input.try(|i| Length::parse(context, i)) {
-                            lengths[length_parsed_count] = value
-                        } else {
-                            break
+                            lengths[3] = value;
```
Comment (upsuper) [inline components/style/values/specified/mod.rs]: You should not try parsing it if `disable_spread_and_inset` is true, since otherwise you may incorrectly accept invalid value.

File: components/style/values/specified/mod.rs [hunk 3]
```diff
@@ -621,19 +621,15 @@ impl Shadow {
...
                 if let Ok(value) = input.try(|i| Length::parse(context, i)) {
                     lengths[0] = value;
-                    let mut length_parsed_count = 1;
-                    while length_parsed_count < length_count {
+
+                    lengths[1] = try!(Length::parse(context, input));
+
+                    if let Ok(value) = input.try(|i| Length::parse_non_negative(i)) {
+                        lengths[2] = value;
+
```
Comment (upsuper) [inline components/style/values/specified/mod.rs]: I wonder do we really need so many empty lines... Could you remove all the three empty lines you add above?

**Review by upsuper — CHANGES_REQUESTED** [review]
In addition, please add a test for this. As I mentioned in #15490, a unit test for `text-shadow` is probably enough.

File: components/style/values/specified/mod.rs (hunk 2 above)
Comment (karan1276) [inline components/style/values/specified/mod.rs]: I could addd an if statement but I think the line 654 compensates for that.

Comment (upsuper) [inline components/style/values/specified/mod.rs]: That makes it not return non-zero spread in that case, but doesn't block the function from parsing invalid value.

**Comment by karan1276** [comment]
I had some concerns about  #15664. Both PRs are going to clash. In fact they might just be the same thing.

**Comment by karan1276** [comment]
Yes ill add some tests

**Comment by upsuper** [comment]
It doesn't seem to me you two are going to clash. You are working on different files, no?

**Comment by karan1276** [comment]
Oh my bad. You are right :)

**Comment by karan1276** [comment]
Hey @upsuper , can you tell me how to go about writing tests?

**Comment by karan1276** [comment]
I was following [this guide](https://github.com/servo/servo/wiki/Property-hacking-guide) . But `tests/unit/style/parsing` files don't look like that. I guess ill have to write a function there. But can you give some more details.

**Comment by jdm** [comment]
The guide's description of writing tests looks like it should apply. Take a look at files like `tests/unit/style/parsing/animation.rs` or `tests/unit/style/parsing/border.rs` (eg. border_image_outset_should_error_on_negative_length).

**Comment by karan1276** [comment]
Thanks, i guess i need to create a new file for box-shadow? Do i need to call that file else where or all the tests are run automatically ?

**Comment by karan1276** [comment]
So some thing is surely wrong because the test always passes. Please help

**Comment by KiChjang** [comment]
It doesn't look like you've written any tests for this PR yet. Have you pushed your changes to your branch?
Also, please remove your merge commit. If you want to update your branch to the latest master, use rebase instead of merge.

**Comment by karan1276** [comment]
aah, i had forgotten to push. Did a rebase.

**Comment by canaltinova** [comment]
Rustc doesn't see the `box_shadow.rs` file because it's not specified as a module right now. You should add box_shadow to [mod.rs](https://dxr.mozilla.org/servo/rev/26de7c6bc48affbc2087b32649850f0733e567f0/tests/unit/style/parsing/mod.rs#80-96).

**Comment by karan1276** [comment]
@canaltinova thanks, ill include that.

File: tests/unit/style/parsing/box_shadow.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use style::properties::longhands::box_shadow;
+
+#[test]
+fn blur_radius_should_not_accept_negavite_values() {
+    let result = parse_longhand!(box_shadow, "normal");
+    assert_eq!(result, "");
```
Comment (upsuper) [inline tests/unit/style/parsing/box_shadow.rs]: This test doesn't look relevant to this change at all. You should test that when the third value is negative, the parse would fail.

File: tests/unit/style/parsing/mod.rs [hunk 5]
```diff
@@ -81,6 +81,7 @@ mod animation;
 mod background;
 mod basic_shape;
 mod border;
+mod box_shadow;
```
Comment (upsuper) [inline tests/unit/style/parsing/mod.rs]: It doesn't seem to me you need a new mod for this test at all. Could you just add the test into `effects.rs`?

You may also add a test to `inherited_text.rs` for `text-shadow`.

**Review by upsuper — CHANGES_REQUESTED** [review]

**Comment by karan1276** [comment]
Can i get help with this: https://gist.github.com/karan1276/61916a46b1d83a1963b0fec5dac58391 thanks

**Comment by upsuper** [comment]
`parse_longhand!` asserts that the parsing succeeds, so you probably cannot use it here. You probably should use
```rust
assert!(parse(box_shadow::parse, "1px 1px -1px").is_err());
```
similiar to https://github.com/servo/servo/blob/d8a8e3e42eda3704d8793d398b24f2e37bded4ec/tests/unit/style/parsing/position.rs#L33-L35

**Comment by upsuper** [comment]
And for the specific issue you meet, the error has told that the value returned from `parse_longhand!` is `box_shadow::SpecifiedValue`, and that cannot be compared with a string.

**Comment by upsuper** [comment]
It seems you need to rebase your patches on top of the master. There is a conflict.

**Comment by Wafflespeanut** [comment]
Please remove the merge commit

**Comment by karan1276** [comment]
Ahh, I pressed "resolve conflicts" on github and it merged. ill rebase right away

**Comment by karan1276** [comment]
Please review, when everything is alright ill squash all the commits.

File: tests/unit/style/parsing/effects.rs [hunk 6]
```diff
@@ -106,3 +107,8 @@ fn test_parse_factor() {
     assert!(parse(filter::parse, "sepia(-1)").is_err());
     assert!(parse(filter::parse, "saturate(-1)").is_err());
 }
+
+fn blur_radius_should_not_accept_negavite_values() {
+    use style::properties::longhands::box_shadow;
+    assert!(parse(box_shadow::parse, "1px 1px -1px").is_err());
```
Comment (upsuper) [inline tests/unit/style/parsing/effects.rs]: Please add a test to ensure that it does parse when the value is valid, e.g. `1px 1px 0px` and `1px 1px 1px`, so that it can be more apparently that the value is rejected because of the negative value, not anything else.

File: components/style/values/specified/mod.rs [hunk 7]
```diff
@@ -655,21 +654,15 @@ impl Shadow {
...
-                    let mut length_parsed_count = 1;
-                    while length_parsed_count < length_count {
-                        if let Ok(value) = input.try(|i| Length::parse(context, i)) {
-                            lengths[length_parsed_count] = value
-                        } else {
-                            break
+                    lengths[1] = try!(Length::parse(context, input));
+                    if let Ok(value) = input.try(|i| Length::parse_non_negative(i)) {
+                        lengths[2] = value;
+                        if !disable_spread_and_inset {
```
Comment (upsuper) [inline components/style/values/specified/mod.rs]: I think you can remove the extra check of `disable_spread_and_inset` at the end of this function returning `Shadow`, since `lengths[3]` would always be `Length::zero()` if that option is `true`. Maybe you can add a `debug_assert!` to test that invariant.

**Review by upsuper — COMMENTED** [review]
It mostly looks good. Please address the issues.

File: components/style/values/specified/mod.rs (hunk 7 above)
Comment (karan1276) [inline components/style/values/specified/mod.rs]: I don't remember exactly but i was talking to someone about `disable_spread_and_inset`, i was told i should not even attempt to parse the spread radius if the option is set, that's why i added that case. If that is not true let me know, ill remove it.

Comment (upsuper) [inline components/style/values/specified/mod.rs]: That's right that you should not attempt to parse the spread radius if the option is set. I'm not asking you to remove the check here, but the one at the end of the function. GitHub doesn't allow me to comment on unchanged lines, so I add comment at this line. Sorry if that made you confused.

Comment (karan1276) [inline components/style/values/specified/mod.rs]: I am not sure about how `debug_assert` works, could you tell me more about it. Or point me to some kind of docs.

Looks like it works like `./mach run --dev --debug`, but i may be wrong

Comment (upsuper) [inline components/style/values/specified/mod.rs]: You can add something like
```rust
debug_assert!(!disable_spread_and_inset || lengths[3] == Length::zero());
```
before returning from the function.

This assertion would run in debug build, so that if this invariant gets broken, we can know it easier without diagnosing too much.

Comment (karan1276) [inline components/style/values/specified/mod.rs]: About `disable_spread_and_inset` that makes sense, ill make the change, cool `debug_assert!` should be a good check.

**Comment by Wafflespeanut** [comment]
Also needs proper squashing before landing.

**Comment by karan1276** [comment]
test error was because of something else, rebase fixed it. All commits are now squashed

**Comment by emilio** [comment]
@bors-servo retry 

 * git exception

**Comment by KiChjang** [comment]
@bors-servo retry
* Exception again
